cclib/bridge/cclib2ase.py

```diff
@@ -0,0 +1,238 @@
+    from ase.calculators.calculator import PropertyNotImplementedError
+
+
+def _check_ase(found_ase):
+    if not found_ase:
+        raise ImportError("You must install `ase` to use this function")
+
+
+def makease(
+    atomcoords, atomnos, atomcharges=None, atomspins=None, atommasses=None
+):
+    """Create an ASE Atoms object from cclib attributes.
+
+    ASE requires atomic partial charges and atomic spin densities rather than
+    molecular charge and multiplicity, so we follow how other interfaces have
+    done (e.g., MOPAC, Gaussian and XTB) and require atomcharges and atomspins,
+    or leave undefined.
+
+    Inputs:
+        atomcoords - two dimensional array-like with atomic coordinates.
+        atomnos - one dimensional array-like with atomic numbers.
+        atomcharges - one dimensional array-like with atomic charges.
+        atomspins - one dimensional array-like with atomic spin densities.
+        atommasses - one dimensional array-like with atomic masses.
+    """
+    _check_ase(_found_ase)
+    return Atoms(
+        positions=atomcoords,
+        numbers=atomnos,
+        masses=atommasses,
+        charges=atomcharges,
+        magmoms=atomspins,
+    )
+
+
+def writetraj(filename, ccdata, popname="mulliken"):
+    """Write an ASE Trajectory object from a ccData object.
+
+    We try to write the following properties: atomcoords, atomnos, atomcharges,
+    atomspins, atommasses, scfenergies, grads, moments (dipole only) and
+    freeenergy. No charge or mult is written since ASE calculates them from
+    atomcharges and atomspins.
+
+    When a program do a final single point calculation at the end of an
+    optimization (e.g., ORCA), we'll have unequal number of grads and
+    atomcoords. This means the last geometry in the written .traj will lack
+    forces.
+
+    Some unit conversions are done here, but things are converted back in
+    readtraj/makecclib.
+
+    Inputs:
+        filename - path to traj file to be written.
+        ccdata - an instance of ccData.
+        popname - population analysis to use for atomic partial charges and
+            atomic spin densities. Molecular charge and multiplicity are
+            evaluated from them.
+    """
+    _check_ase(_found_ase)
+    traj = Trajectory(filename, "w")
+    for i, atomcoords in enumerate(ccdata.atomcoords):
+        atomspins = None
+        if hasattr(ccdata, "atomspins"):
+            atomspins = ccdata.atomspins[popname]
+        atoms = makease(
+            atomcoords,
+            ccdata.atomnos,
+            ccdata.atomcharges[popname],
+            atomspins,
+            ccdata.atommasses,
+        )
+
+        properties = {}
+        if hasattr(ccdata, "scfenergies"):
+            properties.update({"energy": ccdata.scfenergies[i]})
+        if hasattr(ccdata, "grads"):
+            try:
+                properties.update(
+                    {"forces": -ccdata.grads[i] * units.Hartree / units.Bohr}

```

Comment:
This unit conversion is in line with [the one done in the ORCA calculator](https://gitlab.com/ase/ase/-/blob/master/ase/calculators/orca.py#L123).